### DISCUSSION TITLE
Automation between accessories controlled by one ESP32?

### DISCUSSION THREAD

**Discussion by carrefinho** [HomeSpan/HomeSpan#335] [Q&A]
Hi all, I've just got my HomeSpan closet light set up and it's been working amazingly well! However I'd like to add a contact sensor to it and automate the light turning on when the closet door opens. Would it be possible to add that as another accessory and have it somehow control the light accessory locally on the ESP32? I understand that I can implement custom methods but I'm not sure how to correctly reference other accessories and call said methods.

**Comment by HomeSpan** [comment]
Yes, you can do this directly in the same ESP32 device, though there are numerous approaches you can consider.

The easiest is to utilize the HomeSpan `setVal()` function inside the `loop()` method of your current Accessory.  See [Example 12](https://github.com/HomeSpan/HomeSpan/tree/master/examples/12-ServiceLoops) for details on how to do this.  From within the `loop()` method you'll need to create some logic to read a pin that is connected to your contact sensor.  When the pin shows the sensor has gone from closed to open, you can trigger the relay controlling the light to turn on, as well as use `setVal()` to update the Home App indicating the light is now on.  Similarly, you'll want to trigger the light to turn off when the contact goes from open to closed, as well as use `setVal()` to update the Home App indicating the light is now off.  You'll also need to write some logic to "debounce" the contact sensor switch (I usually use a simple exponential decay that "slowly" increases/decreases a variable when the pin is either high or low).

A second way to do this would be to create a separate Contact Sensor Service (see the HAP-R2 guide for details) from either within the same Accessory, or by instantiating a second Accessory.   Then, rather than create any logic in HomeSpan, I believe you could use Automation routines in the Home App itself to trigger actions (such as turning on the closet light) when the Contact Sensor changes states.  I've previously tested this with a StatelessProgrammableSwitch, but have not used Home App automations with Contact Sensors, so not 100% sure how this would work in the Home App.

Note the second method is required if your contact switch and light control electronics are connected to  _different_ ESP32 devices. If everything is on the same device, either method should be applicable, though I prefer the first since it avoids needing to have HomeKit running on your network to process the Automation routines.
